### PR TITLE
Pattern-match callbacks

### PR DESCRIPTION
**Pull request by waralex**
### Pattern-match callbacks. 

The functionality is the same as in Python, with a single nuance - NamedTuple is used as the id:

```julia
callback!(app,
    Output((type = "output", index = MATCH), "children"),
    Input((type = "dropdown", index = MATCH), "value"),
    State((type = "dropdown", index = MATCH), "id"),
) do value, id
    return html_div("Dropdown $(id.index) = $(value)")
end
```

### CODE REVIEW

File: src/app/supporttypes.jl [hunk 1]
```diff
@@ -1,28 +1,97 @@
...
+function Base.:(==)(a::Dependency, b::Dependency) 
+    (a.property == b.property) && is_id_matches(a, b)
+end
+
+function Base.isequal(a::Dependency, b::Dependency)
+    return a == b
+end
+
+#The regular unique works using Set and will not work correctly, because in our case "equal" dependencies have different hashes
+#This implementation is not algorithmically efficient (O(n^2)), but it is quite suitable for checking the uniqueness of outputs
```
Comment (alexcjohnson) [inline src/app/supporttypes.jl:49]: Yep, that should be fine - this only happens when starting the server anyway, we only really care about efficiency during callback requests. Readability and simplicity are more important here.

File: test/callbacks.jl [hunk 2]
```diff
@@ -37,8 +37,8 @@ using Dash
     response = HTTP.handle(handler, request)
     @test response.status == 200
     resp_obj = JSON2.read(String(response.body))
-    @test !in(:multi, keys(resp_obj))
-    @test resp_obj.response.props.children == "test"
+    @test in(:multi, keys(resp_obj))
```
Comment (alexcjohnson) [inline test/callbacks.jl:40]: FYI I see a few trailing whitepaces added in this PR (and a few others removed) - it's a little easier to review if we never have these, do you use an editor that can be set to automatically strip these on saving?

**Comment by alexcjohnson** [comment]
Is the resource refactoring here solving a problem, or just simplifying the structure?

**Comment by waralex** [comment]
@alexcjohnson
>Is the resource refactoring here solving a problem, or just simplifying the structure?

Initially, I took resources from the R version, where they were grouped into folders. Now it turned out that this version of resources does not support pattern-matching, so I updated the resources to those used in Python dash 1.4.3. Python dash does not split them into directories, and it seemed to me that this is more convenient.

**Comment by waralex** [comment]
@alexcjohnson 
> It looks like this is handled self-consistently, meaning that when you get a pattern-matching ID in callback_context it's also a NamedTuple?

It depends on which context parameter we are talking about. `inputs_list`, `states_list`, and `outputs_list`- Yes, id is NamedTuple. In `triggered` and  the `inputs` and  the `states` dictionaries, the keys are like in Python-strings from json with sorted keys + `.<property>`. On the one hand, for consistency with the Python api, on the other, because in this form they come to changedProps from the front end:
```
triggered : [(prop_id = "{\"action\":\"done\",\"item\":3}.value", value = Any["done"])]
inputs:  Dict("{\"action\":\"done\",\"item\":3}.value" => Any["done"])
inputs_list: [(id = (action = "done", item = 3), property = "value", value = Any["done"])]
outputs_list: [(id = (item = 3,), property = "style")]
```

In principle, there is no big problem to make the keys also NamedTuple. Just keep in mind that the order of fields is important in NamedTuple, i.e. `(a=1, b=2) != (b=2, a=1)`

**Comment by alexcjohnson** [comment]
> On the one hand, for consistency with the Python api, on the other, because in this form they come to changedProps from the front end

Perfect. 💃 still applies.

**Pull request merged**